Mathlib/Analysis/LocallyConvex/Bipolar.lean

```diff
@@ -0,0 +1,166 @@
+- `LinearMap.rightDualEquiv`: When `B` is right-separating, `F` is linearly equivalent to the
+  topological dual of `E` with the weak topology.
+- `LinearMap.leftDualEquiv`: When `B` is left-separating, `E` is linearly equivalent to the
+  topological dual of `F` with the weak topology.
+
+## Main statements
+
+- `LinearMap.flip_polar_polar_eq`: The Bipolar Theorem: The bipolar of a set coincides with its
+  closed absolutely convex hull.
+
+## References
+
+* [Conway, *A course in functional analysis*][conway1990]
+
+## Tags
+
+bipolar
+-/
+
+variable {𝕜 E F : Type*}
+
+namespace LinearMap
+
+section NormedField
+
+variable {𝕜 E F : Type*}
+variable [NormedField 𝕜] [NormedSpace ℝ 𝕜] [AddCommMonoid E] [AddCommMonoid F]
+variable [Module 𝕜 E] [Module 𝕜 F]
+
+variable {B : E →ₗ[𝕜] F →ₗ[𝕜] 𝕜} (s : Set E)
+
+variable [Module ℝ F] [IsScalarTower ℝ 𝕜 F] [IsScalarTower ℝ 𝕜 𝕜]
+
+theorem polar_AbsConvex : AbsConvex 𝕜 (B.polar s) := by
+  rw [polar_eq_biInter_preimage]
+  exact AbsConvex.iInter₂ fun i hi =>
+    ⟨balanced_closedBall_zero.mulActionHom_preimage (f := (B i : (F →ₑ[(RingHom.id 𝕜)] 𝕜))),
+      (convex_closedBall _ _).linear_preimage (B i)⟩
+
+end NormedField
+
+variable [RCLike 𝕜] [AddCommGroup E] [AddCommGroup F]
+variable [Module 𝕜 E] [Module 𝕜 F]
+
+variable (B : E →ₗ[𝕜] F →ₗ[𝕜] 𝕜)
+
+lemma isBounded_of_Continuous (f : WeakBilin B →L[𝕜] 𝕜) :
+    Seminorm.IsBounded B.toSeminormFamily (fun _ : Fin 1 => normSeminorm 𝕜 𝕜) f.toLinearMap := by
+  obtain ⟨s,C, hC1, hC2⟩ :=
+    Seminorm.bound_of_continuous B.weakBilin_withSeminorms
+      f.toSeminorm (continuous_norm.comp f.continuous)
+  rw [Seminorm.IsBounded, forall_const]
+  exact ⟨s, ⟨C, hC2⟩⟩
+
+lemma dualEmbedding_isSurjective : Function.Surjective (WeakBilin.eval B) := by
+  intro f₁
+  have test5 : ∃ (s₁ : Finset F),
+    ↑f₁ ∈ Submodule.span 𝕜 (Set.range (ContinuousLinearMap.toLinearMap₁₂
+      (WeakBilin.eval B) ∘ Subtype.val : s₁ → WeakBilin B →ₗ[𝕜] 𝕜)) := by
+    obtain ⟨s,hS⟩ := isBounded_of_Continuous B f₁ (Fin.last 0)
+    exact ⟨s, functional_mem_span_iff.mpr hS⟩
+  obtain ⟨s, hs⟩ := test5
+  rw [← Set.image_univ, Finsupp.mem_span_image_iff_linearCombination] at hs
+  obtain ⟨l, _, hl2⟩ := hs
+  use Finsupp.linearCombination 𝕜 Subtype.val l
+  rw [←ContinuousLinearMap.coe_inj, ← hl2, WeakBilin.eval, coe_mk, AddHom.coe_mk]
+  simp only
+  rw [ContinuousLinearMap.toLinearMap₁₂, ContinuousLinearMap.coeLMₛₗ,
+    Finsupp.linearCombination_apply, Finsupp.linearCombination_apply, map_finsuppSum]
+  simp
+
+lemma dualEmbedding_isInjective_of_separatingRight (hr : B.SeparatingRight) :
+    Function.Injective (WeakBilin.eval B) := (injective_iff_map_eq_zero _).mpr (fun f hf =>
+    (separatingRight_iff_linear_flip_nontrivial.mp hr) f (ContinuousLinearMap.coe_inj.mpr hf))
+
+/-- When `B` is right-separating, `F` is linearly equivalent to the topological dual of `E` with the
+weak topology. -/
+noncomputable def rightDualEquiv (hr : B.SeparatingRight) : F ≃ₗ[𝕜] (WeakBilin B) →L[𝕜] 𝕜 :=
```

Comment:
```suggestion
noncomputable def rightDualEquiv (hr : B.SeparatingRight) : F ≃ₗ[𝕜] NormedSpace.Dual 𝕜 (WeakBilin B) :=
```

Reply:
i.e. we don't have `SeminormedAddCommGroup E`?

Reply:
Hmn I guess this is an aspirational suggestion. `NormedSpace.Dual` is just an abbrev, so we could relax its typeclass assumptions without any issue (in a separate PR)

Reply:
@eric-wieser how about this: #27699 - probably more general than needed?

Reply:
I've left a comment on that PR. There's a natural name for `E →L[𝕜] 𝕜` with the topology we already have on that space: it's the *strong dual*.